src/write-archive/archive-file.ts

```diff
@@ -0,0 +1,88 @@
+import mime from 'mime';
+import path from 'path';
+import { createHash } from 'node:crypto';
+import { logger } from '../utils/logger';
+import type { ArchiveResponse, UrlString } from '../resource-archive';
+
+export class ArchiveFile {
+  url: URL;
+
+  response: ArchiveResponse;
+
+  shortenedFileNameLength: number;
+
+  constructor(url: UrlString, response: ArchiveResponse) {
+    this.url = new URL(url);
+    this.response = response;
+    this.shortenedFileNameLength = 250;
+  }
+
+  originalSrc() {
+    // Assets that we capture will be stripped of the domain in the source
+    return `${this.url.pathname}${this.url.search}`;
+  }
+
+  toFileSystemPath() {
+    let sanitizedSrc = this.url.pathname;
+
+    sanitizedSrc = this.ensureNonDirectory(sanitizedSrc);
+    sanitizedSrc = this.encodeQueryString(sanitizedSrc);
+    sanitizedSrc = this.truncateFileName(sanitizedSrc);
+    sanitizedSrc = this.addExtension(sanitizedSrc);
+
+    return sanitizedSrc;
+  }
+
+  private ensureNonDirectory(pathname: string) {
+    // TODO maybe drop the .html?
+    return pathname.endsWith('/') ? `${pathname}index.html` : pathname;
+  }
+
+  private encodeQueryString(pathname: string) {
+    const queryString = this.url.search;
+    if (!queryString) return pathname;
+
+    // TODO maybe prepend the last file name in case it has an extension?
+    const safeQueryString = this.hash(queryString);
```

Comment:
I'm hashing the query string instead of stripping special chars because it could be long. If it's a long URL to specify which image should be loaded, for example, there's a risk of the image name that differentiates it from the other ones will get truncated, leading to a new problem with collisions.

Reply:
Oh, I didn't even think of this case. Good catch!

Reply:
You know, this makes me wonder if perhaps we should hash _all_ the filename values. If we were to output an md5 hash with a type extension for a filename, instead of the custom logic we do now, we'd be guaranteed that there wouldn't be anything over the character limit. Although, it's probably best to leave it as-is for now.

Reply:
Yeah maybe! It might be nice to leave things that don't need to change. But it's probably safer to hash whenever we do need to change the path, instead of truncating, etc.